clients/da-vinci-client/src/main/java/com/linkedin/davinci/validation/PartitionTracker.java

```diff
@@ -632,7 +691,43 @@ public void checkMissingMessage(
           // no-op
       }
     } finally {
-      partitionLock.unlock();
+      lockAndSegment.unlock();
+    }
+  }
+
+  void clearExpiredStateAndUpdateOffsetRecord(OffsetRecord offsetRecord, long maxAgeInMs) {
+    long minimumRequiredRecordProducerTimestamp = offsetRecord.getMaxMessageTimeInMs() - maxAgeInMs;
+    int numberOfClearedGUIDs = 0;
+    Iterator<Map.Entry<GUID, ReentrantLockWithRef<Segment>>> iterator = this.segments.entrySet().iterator();
+    Map.Entry<GUID, ReentrantLockWithRef<Segment>> entry;
+    ReentrantLockWithRef<Segment> lockAndSegment;
+    while (iterator.hasNext()) {
+      entry = iterator.next();
+      lockAndSegment = entry.getValue();
+      if (!lockAndSegment.containsRef()) {
+        // lock-free check
+        continue;
+      }
+      lockAndSegment.lock();
+      try {
+        if (!lockAndSegment.containsRef()) {
+          // lock-ful check
+          continue;
+        }
+        if (lockAndSegment.getRef().getLastRecordProducerTimestamp() < minimumRequiredRecordProducerTimestamp) {
+          lockAndSegment.removeRef();
+          iterator.remove();
+          offsetRecord.removeProducerPartitionState(entry.getKey());
+          numberOfClearedGUIDs++;
+        } else {
+          updateOffsetRecord(entry.getKey(), lockAndSegment, offsetRecord);
+        }
+      } finally {
+        lockAndSegment.unlock();
+      }
+    }
+    if (numberOfClearedGUIDs > 0) {
+      logger.info("Cleared {} expired producer GUID(s).", numberOfClearedGUIDs);
```

Comment:
Can we also put `partition id` and `topic name` into the log message?

Reply:
It will be there. The `logger` is a class property, which is prefixed by topic name and partition id. See the constructor.